logins-sql/src/update_plan.rs

```diff
@@ -0,0 +1,252 @@
+                (":username_field",  &login.username_field as &ToSql),
+                (":password_field",  &login.password_field as &ToSql),
+                (":password",        &login.password as &ToSql),
+                (":hostname",        &login.hostname as &ToSql),
+                (":username",        &login.username as &ToSql),
+
+                (":times_used",            &login.times_used as &ToSql),
+                (":time_last_used",        &login.time_last_used as &ToSql),
+                (":time_password_changed", &login.time_password_changed as &ToSql),
+                (":time_created",          &login.time_created as &ToSql),
+
+                (":guid", &login.guid_str() as &ToSql),
+            ])?;
+        }
+        Ok(())
+    }
+
+    fn perform_local_updates(&self, tx: &mut Transaction) -> Result<()> {
+        let sql = format!("
+            UPDATE loginsL
+            SET local_modified      = :local_modified,
+                httpRealm           = :http_realm,
+                formSubmitURL       = :form_submit_url,
+                usernameField       = :username_field,
+                passwordField       = :password_field,
+                timeLastUsed        = :time_last_used,
+                timePasswordChanged = :time_password_changed,
+                timesUsed           = :times_used,
+                password            = :password,
+                hostname            = :hostname,
+                username            = :username,
+                sync_status         = {changed}
+            WHERE guid = :guid",
+            changed = SyncStatus::Changed as u8);
+        let mut stmt = tx.prepare_cached(&sql)?;
+        // XXX OutgoingChangeset should no longer have timestamp.
+        let local_ms: i64 = util::system_time_ms_i64(SystemTime::now());
+        for l in &self.local_updates {
+            trace!("Updating local {:?}", l.guid_str());
+            stmt.execute_named(&[
+                (":local_modified", &local_ms as &ToSql),
+
+                (":http_realm",      &l.login.http_realm as &ToSql),
+                (":form_submit_url", &l.login.form_submit_url as &ToSql),
+                (":username_field",  &l.login.username_field as &ToSql),
+                (":password_field",  &l.login.password_field as &ToSql),
+                (":password",        &l.login.password as &ToSql),
+                (":hostname",        &l.login.hostname as &ToSql),
+                (":username",        &l.login.username as &ToSql),
+
+                (":time_last_used",        &l.login.time_last_used as &ToSql),
+                (":time_password_changed", &l.login.time_password_changed as &ToSql),
+                (":times_used",            &l.login.times_used as &ToSql),
+
+                (":guid", &l.guid_str() as &ToSql),
+            ])?;
+        }
+        Ok(())
+    }
+
+    pub fn execute(&self, tx: &mut Transaction) -> Result<()> {
```

Comment:
This is actually pretty awesome.  I like it a lot!

Reply:
Everything is better when structured as a command buffer.